### PR TITLE
[Sources] Update file search results when source changes

### PR DESCRIPTION
**Pull request by kenjiO**
Fixes Issue: #6544

Can you let me know if I am on the right track by doing it this way?

### CODE REVIEW

File: src/actions/sources/select.js [hunk 1]
```diff
@@ -137,6 +141,19 @@ export function selectLocation(location: Location) {
...
+    const isLocationChanged =
+      !previousLocation || previousLocation.sourceId !== location.sourceId;
+    if (isLocationChanged) {
+      const isFileSearchOpen = getActiveSearch(getState()) === "file";
+      const fileSearchQuery = getFileSearchQuery(getState());
+      if (isFileSearchOpen && fileSearchQuery) {
+        const editor = getEditor();
+        dispatch(searchContents(fileSearchQuery, editor));
+      }
+    }
```
Comment (jasonLaster) [inline src/actions/sources/select.js]: lets try and simplify this a bit...

```js
const newSource = getSelectedSource(getState());
if (currentSource != newSource) {
  dispatch(clearFileSearch())
}
```

we can do a couple of things in `clearFileSearch`

1. check if there is an active file search
2. dispatch `searchContents`

File: src/actions/sources/select.js [hunk 2]
```diff
@@ -142,17 +144,10 @@ export function selectLocation(location: Location) {
...
-    if (isLocationChanged) {
-      const isFileSearchOpen = getActiveSearch(getState()) === "file";
-      const fileSearchQuery = getFileSearchQuery(getState());
-      if (isFileSearchOpen && fileSearchQuery) {
-        const editor = getEditor();
-        dispatch(searchContents(fileSearchQuery, editor));
-      }
+    // If a new source is selected update the file search results
+    const newSource = getSelectedSource(getState());
+    if (!currentSource || (newSource && currentSource.id != newSource.id)) {
```
Comment (kenjiO) [inline src/actions/sources/select.js]: I tried doing `(currentSource != newSource)` to check if the source was changed but that was evaluating to true even when the source wasn't changed.  So I used this instead

**Comment by kenjiO** [comment]
I noticed `selectLocation` and `selectSpecificLocation` are the same except one has an if statement that does some prettyPrint actions.  Do you want to refactor this so there is not so much repeated code?

**Comment by jasonLaster** [comment]
Take a stab. :)

**Comment by kenjiO** [comment]
How does this look?

File: src/actions/sources/select.js [hunk 3]
```diff
@@ -96,87 +100,15 @@ export function selectSource(sourceId: string) {
...
-      isMinified(loadedSource)
-    ) {
-      await dispatch(togglePrettyPrint(loadedSource.id));
-      dispatch(closeTab(loadedSource.url));
-    }
-
-    dispatch(setSymbols(loadedSource.id));
-    dispatch(setOutOfScopeLocations());
-  };
+  return locationSelect(location, {checkPrettyPrint: true});
```
Comment (jasonLaster) [inline src/actions/sources/select.js]: looks good, maybe instead of `locatonSelect`, we add a param  here `selectLocation(location: Location, {checkPrettyPrint: true})`

also lets do this in a separate PR

**Comment by kenjiO** [comment]
I took out the refactor and will do that in a separate PR.  Is there anything else that I should change for this PR?

File: src/actions/sources/select.js [hunk 4]
```diff
@@ -137,6 +143,12 @@ export function selectLocation(location: Location) {
 
     dispatch(setSymbols(loadedSource.id));
     dispatch(setOutOfScopeLocations());
+
+    // If a new source is selected update the file search results
+    const newSource = getSelectedSource(getState());
+    if (!currentSource || (newSource && currentSource.id != newSource.id)) {
```
Comment (jasonLaster) [inline src/actions/sources/select.js]: maybe we can be even simpler. This will say, there had to have been a current source and it can't be the same as the new one. I don't know if the IDs matter

```js
if (currentSource && currentSource !== newSource) {
}
```

File: src/actions/sources/select.js [hunk 5]
```diff
@@ -227,3 +247,17 @@ export function jumpToMappedSelectedLocation() {
...
+function clearFileSearch() {
+  return async ({ dispatch, getState }: ThunkArgs) => {
+    const isFileSearchOpen = getActiveSearch(getState()) === "file";
+    const fileSearchQuery = getFileSearchQuery(getState());
+    if (isFileSearchOpen && fileSearchQuery) {
+      const editor = getEditor();
+      dispatch(searchContents(fileSearchQuery, editor));
+    }
+  };
+}
```
Comment (jasonLaster) [inline src/actions/sources/select.js]: lets move this to a different action file that is focused on search

File: src/actions/sources/select.js (hunk 4 above)
Comment (kenjiO) [reply src/actions/sources/select.js]: I tried it this way but it would sometimes return true even when the source did not change, such as when clicking on the editor tab or source file of the file that is already active.  The objects were different even though they were pointing to the same source.
screen shot 2018-07-05 at 10 44 21 am

File: src/actions/sources/select.js (hunk 5 above)
Comment (kenjiO) [reply src/actions/sources/select.js]: I moved it to the ui.js actions file because that has the similar actions `closeActiveSearch` and `setActiveSearch`.  But it if you prefer it being in the file-seach.js actions file I could move it there instead.  I also renamed the action to `updateActiveFileSeach` since I think that better describes what it does rather than `clearFileSearch`

File: src/actions/sources/select.js (hunk 4 above)
Comment (jasonLaster) [reply src/actions/sources/select.js]: nice find. try changing this and see if that is still the case. If it is, lets figure out why the object is changing!

```diff

diff --git a/src/reducers/sources.js b/src/reducers/sources.js
index bb18ebc..a74bf7c 100644
--- a/src/reducers/sources.js
+++ b/src/reducers/sources.js
@@ -416,15 +416,7 @@ export const getSelectedSource = createSelector(
       return;
     }

-    const source = sources[selectedLocation.sourceId];
-
-    // TODO: remove this when the immutable refactor lands in m-c
-    if (isTesting) {
-      const testSource: any = { ...source, get: field => source[field] };
-      return testSource;
-    }
-
-    return source;
+    return sources[selectedLocation.sourceId];
   }
 );
```

Comment (jasonLaster) [reply src/actions/sources/select.js]: just noticed this should be ` if (isTesting()) {`

Comment (kenjiO) [reply src/actions/sources/select.js]: I added the parentheses and that fixed it and now comparing for the same object works :)

**Comment by jasonLaster** [comment]
Wow! I love how clean this PR became and how we uncovered a bug from the immutable refactoring work
